<a href="https://colab.research.google.com/github/ZiyangZ/Machine-learning/blob/master/HW2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#Load data

Use 80% data as training data and 20% as testing data.

In [3]:
from keras.datasets import cifar10

(train_images_original, train_labels_original), (test_images_original, test_labels_original) = cifar10.load_data()

train_images_all = train_images_original.astype('float32') / 255
test_images = test_images_original.astype('float32') / 255

from keras.utils import to_categorical

train_labels_all = to_categorical(train_labels_original)
test_labels = to_categorical(test_labels_original)

train_images = train_images_all[:40000,:]
validation_images = train_images_all[40000:,:]
train_labels = train_labels_all[:40000,:]
validation_labels = train_labels_all[40000:,:]

Using TensorFlow backend.


In [4]:
print(train_images.shape, train_labels.shape, validation_images.shape, validation_labels.shape, test_images.shape, test_labels.shape)

(40000, 32, 32, 3) (40000, 10) (10000, 32, 32, 3) (10000, 10) (10000, 32, 32, 3) (10000, 10)


#Architecture 1

Basic architecture.

In [0]:
from keras import layers
from keras import models
from keras import optimizers

batch_size = 32
epochs = 100

model = models.Sequential()

model.add(layers.Conv2D(16, (3, 3), activation='relu', input_shape=(32, 32, 3)))
model.add(layers.MaxPooling2D(2, 2))

model.add(layers.Conv2D(32, (3, 3), activation='relu'))
model.add(layers.MaxPooling2D(2, 2))

model.add(layers.Conv2D(64, (3, 3), activation='relu'))
model.add(layers.MaxPooling2D(2, 2))

model.add(layers.Flatten())
model.add(layers.Dense(512, activation='relu'))
model.add(layers.Dense(10, activation='softmax'))

model.compile(
    loss='categorical_crossentropy', 
    optimizer=optimizers.RMSprop(lr=1e-4),
    metrics=['acc'])

model.fit(train_images, train_labels,
              batch_size=batch_size,
              epochs=epochs,
              validation_data=(validation_images, validation_labels))

Train on 40000 samples, validate on 10000 samples
Epoch 1/100
40000/40000 [==============================] - 12s 298us/step - loss: 1.9809 - acc: 0.2794 - val_loss: 1.8140 - val_acc: 0.3509
Epoch 2/100
40000/40000 [==============================] - 12s 303us/step - loss: 1.7211 - acc: 0.3791 - val_loss: 1.6465 - val_acc: 0.4058
Epoch 3/100
40000/40000 [==============================] - 12s 301us/step - loss: 1.6029 - acc: 0.4231 - val_loss: 1.5777 - val_acc: 0.4353
Epoch 4/100
40000/40000 [==============================] - 12s 307us/step - loss: 1.5226 - acc: 0.4542 - val_loss: 1.4828 - val_acc: 0.4644
Epoch 5/100
40000/40000 [==============================] - 12s 295us/step - loss: 1.4595 - acc: 0.4787 - val_loss: 1.4344 - val_acc: 0.4916
Epoch 6/100
40000/40000 [==============================] - 11s 277us/step - loss: 1.4066 - acc: 0.4990 - val_loss: 1.4098 - val_acc: 0.5057
Epoch 7/100
40000/40000 [==============================] - 11s 276us/step - loss: 1.3628 - acc: 0.5167 - val_l

After 100 epochs, the validation accuracy is 0.6856

In [0]:
del model

#Architecture 2

Modify architecture 1 by adding additional layers.

In [0]:
from keras import layers
from keras import models
from keras import optimizers

batch_size = 32
epochs = 50

model = models.Sequential()

model.add(layers.Conv2D(32, (3, 3), activation='relu', input_shape=(32, 32, 3)))
model.add(layers.Conv2D(32, (3, 3), activation='relu'))
model.add(layers.MaxPooling2D(2, 2))

model.add(layers.Conv2D(64, (3, 3), activation='relu'))
model.add(layers.Conv2D(64, (3, 3), activation='relu'))
model.add(layers.MaxPooling2D(2, 2))

model.add(layers.Flatten())
model.add(layers.Dense(512, activation='relu'))
model.add(layers.Dense(10, activation='softmax'))

model.compile(
    loss='categorical_crossentropy', 
    optimizer=optimizers.RMSprop(lr=1e-4),
    metrics=['acc'])

model.fit(train_images, train_labels,
              batch_size=batch_size,
              epochs=epochs,
              validation_data=(validation_images, validation_labels))
              

Train on 40000 samples, validate on 10000 samples
Epoch 1/50
40000/40000 [==============================] - 17s 435us/step - loss: 1.7354 - acc: 0.3723 - val_loss: 1.4965 - val_acc: 0.4643
Epoch 2/50
40000/40000 [==============================] - 17s 419us/step - loss: 1.4357 - acc: 0.4853 - val_loss: 1.3656 - val_acc: 0.5187
Epoch 3/50
40000/40000 [==============================] - 17s 414us/step - loss: 1.3109 - acc: 0.5357 - val_loss: 1.3714 - val_acc: 0.5137
Epoch 4/50
40000/40000 [==============================] - 17s 413us/step - loss: 1.2144 - acc: 0.5720 - val_loss: 1.2209 - val_acc: 0.5713
Epoch 5/50
40000/40000 [==============================] - 17s 419us/step - loss: 1.1312 - acc: 0.6058 - val_loss: 1.1579 - val_acc: 0.5913
Epoch 6/50
40000/40000 [==============================] - 17s 413us/step - loss: 1.0573 - acc: 0.6319 - val_loss: 1.0920 - val_acc: 0.6208
Epoch 7/50
40000/40000 [==============================] - 16s 412us/step - loss: 0.9962 - acc: 0.6548 - val_loss: 1.

The validation accuracy is slightly increased to 0.7253

In [0]:
del model

#Architecture 3

Add dropout layers to architecture 2.

In [0]:
from keras import layers
from keras import models
from keras import optimizers

batch_size = 32
epochs = 50

model = models.Sequential()

model.add(layers.Conv2D(32, (3, 3), activation='relu', input_shape=(32, 32, 3)))
model.add(layers.Conv2D(32, (3, 3), activation='relu'))
model.add(layers.MaxPooling2D(2, 2))
model.add(layers.Dropout(0.25))

model.add(layers.Conv2D(64, (3, 3), activation='relu'))
model.add(layers.Conv2D(64, (3, 3), activation='relu'))
model.add(layers.MaxPooling2D(2, 2))
model.add(layers.Dropout(0.25))

model.add(layers.Flatten())
model.add(layers.Dense(512, activation='relu'))
model.add(layers.Dropout(0.25))
model.add(layers.Dense(10, activation='softmax'))

model.compile(
    loss='categorical_crossentropy', 
    optimizer=optimizers.RMSprop(lr=1e-4),
    metrics=['acc'])

model.fit(train_images, train_labels,
              batch_size=batch_size,
              epochs=epochs,
              validation_data=(validation_images, validation_labels))

Instructions for updating:
Please use `rate` instead of `keep_prob`. Rate should be set to `rate = 1 - keep_prob`.
Train on 40000 samples, validate on 10000 samples
Epoch 1/50
40000/40000 [==============================] - 19s 482us/step - loss: 1.8610 - acc: 0.3193 - val_loss: 1.6286 - val_acc: 0.4112
Epoch 2/50
40000/40000 [==============================] - 19s 464us/step - loss: 1.5656 - acc: 0.4282 - val_loss: 1.4407 - val_acc: 0.4764
Epoch 3/50
40000/40000 [==============================] - 18s 462us/step - loss: 1.4416 - acc: 0.4785 - val_loss: 1.3539 - val_acc: 0.5184
Epoch 4/50
40000/40000 [==============================] - 19s 465us/step - loss: 1.3528 - acc: 0.5143 - val_loss: 1.2706 - val_acc: 0.5522
Epoch 5/50
40000/40000 [==============================] - 19s 463us/step - loss: 1.2800 - acc: 0.5401 - val_loss: 1.2169 - val_acc: 0.5680
Epoch 6/50
40000/40000 [==============================] - 19s 466us/step - loss: 1.2199 - acc: 0.5696 - val_loss: 1.1411 - val_acc: 0.6021
E

The validation accuracy is increased to 0.7854.

In [0]:
del model

#Architecture 4

Add image data augmentation to architecture 3.

In [5]:
from keras.preprocessing.image import ImageDataGenerator
from keras import layers
from keras import models
from keras import optimizers

batch_size = 32
epochs = 100

model = models.Sequential()

model.add(layers.Conv2D(32, (3, 3), activation='relu', input_shape=(32, 32, 3)))
model.add(layers.Conv2D(32, (3, 3), activation='relu'))
model.add(layers.MaxPooling2D(2, 2))
model.add(layers.Dropout(0.25))

model.add(layers.Conv2D(64, (3, 3), activation='relu'))
model.add(layers.Conv2D(64, (3, 3), activation='relu'))
model.add(layers.MaxPooling2D(2, 2))
model.add(layers.Dropout(0.25))

model.add(layers.Flatten())
model.add(layers.Dense(512, activation='relu'))
model.add(layers.Dropout(0.25))
model.add(layers.Dense(10, activation='softmax'))

model.compile(
    loss='categorical_crossentropy', 
    optimizer=optimizers.RMSprop(lr=1e-4),
    metrics=['acc'])

datagen = ImageDataGenerator(
    rotation_range=1,
    width_shift_range=0.1,
    height_shift_range=0.1,
    horizontal_flip=True,
    fill_mode='nearest')

datagen.fit(train_images)

model.fit_generator(datagen.flow(train_images, train_labels,
              batch_size=batch_size),
              steps_per_epoch = 1250,
              epochs=epochs,
              validation_data=(validation_images, validation_labels))



Instructions for updating:
Colocations handled automatically by placer.
Instructions for updating:
Please use `rate` instead of `keep_prob`. Rate should be set to `rate = 1 - keep_prob`.
Instructions for updating:
Use tf.cast instead.
Epoch 1/100
1250/1250 [==============================] - 32s 26ms/step - loss: 1.8850 - acc: 0.3043 - val_loss: 1.6333 - val_acc: 0.4066
Epoch 2/100
1250/1250 [==============================] - 31s 25ms/step - loss: 1.6007 - acc: 0.4091 - val_loss: 1.4275 - val_acc: 0.4832
Epoch 3/100
1250/1250 [==============================] - 31s 25ms/step - loss: 1.4900 - acc: 0.4596 - val_loss: 1.3677 - val_acc: 0.5094
Epoch 4/100
1250/1250 [==============================] - 31s 25ms/step - loss: 1.4226 - acc: 0.4833 - val_loss: 1.2740 - val_acc: 0.5468
Epoch 5/100
1250/1250 [==============================] - 31s 25ms/step - loss: 1.3610 - acc: 0.5084 - val_loss: 1.2242 - val_acc: 0.5657
Epoch 6/100
1250/1250 [==============================] - 31s 25ms/step - loss: 1

The validation accuracy is slightly decreased to 0.7658.

In [0]:
model.save('cifar10_model4.h5')

#Using architecture 4 to predict the test data

In [8]:
from keras.models import load_model

model = load_model('cifar10_model4.h5')
test_loss, test_acc = model.evaluate(test_images, test_labels)
print(test_loss, test_acc)

10000/10000 [==============================] - 1s 138us/step
0.6950683030128479 0.7608


The prediction accuracy on test data is 0.7608, which is very close to the validation accuracy 0.7658 in architecture 4.

In [0]:
del model

#Using 5-fold cross-validation to fit the training data using Architecture 4

In [20]:
from keras.preprocessing.image import ImageDataGenerator
from keras import layers
from keras import models
from keras import optimizers
from sklearn.model_selection import StratifiedKFold
import numpy as np

batch_size = 32
epochs = 100

seed = 8
np.random.seed(seed)

validate_accuracy = []
test_accuracy = []

kfold = StratifiedKFold(n_splits=5, shuffle=True, random_state=seed)

for train, validate in kfold.split(train_images_all, train_labels_original):
  model = models.Sequential()

  model.add(layers.Conv2D(32, (3, 3), activation='relu', input_shape=(32, 32, 3)))
  model.add(layers.Conv2D(32, (3, 3), activation='relu'))
  model.add(layers.MaxPooling2D(2, 2))
  model.add(layers.Dropout(0.25))

  model.add(layers.Conv2D(64, (3, 3), activation='relu'))
  model.add(layers.Conv2D(64, (3, 3), activation='relu'))
  model.add(layers.MaxPooling2D(2, 2))
  model.add(layers.Dropout(0.25))

  model.add(layers.Flatten())
  model.add(layers.Dense(512, activation='relu'))
  model.add(layers.Dropout(0.25))
  model.add(layers.Dense(10, activation='softmax'))

  model.compile(
    loss='categorical_crossentropy', 
    optimizer=optimizers.RMSprop(lr=1e-4),
    metrics=['acc'])

  datagen = ImageDataGenerator(
    rotation_range=1,
    width_shift_range=0.1,
    height_shift_range=0.1,
    horizontal_flip=True,
    fill_mode='nearest')

  datagen.fit(train_images_all[train])

  model.fit_generator(datagen.flow(train_images_all[train], train_labels_all[train],
              batch_size=batch_size),
              steps_per_epoch = 1250,
              epochs=epochs,
              validation_data=(train_images_all[validate], train_labels_all[validate]))
  
  v_loss, v_acc = model.evaluate(train_images_all[validate], train_labels_all[validate])
  t_loss, t_acc = model.evaluate(test_images, test_labels)
  validate_accuracy.append(v_acc)
  test_accuracy.append(t_acc)

print(validate_accuracy, test_accuracy)
print(np.mean(validate_accuracy), np.mean(test_accuracy))


Epoch 1/100
1250/1250 [==============================] - 32s 25ms/step - loss: 1.8888 - acc: 0.3008 - val_loss: 1.6476 - val_acc: 0.3975
Epoch 2/100
1250/1250 [==============================] - 31s 25ms/step - loss: 1.6223 - acc: 0.4024 - val_loss: 1.4664 - val_acc: 0.4593
Epoch 3/100
1250/1250 [==============================] - 31s 25ms/step - loss: 1.5065 - acc: 0.4480 - val_loss: 1.3893 - val_acc: 0.4847
Epoch 4/100
1250/1250 [==============================] - 31s 25ms/step - loss: 1.4360 - acc: 0.4797 - val_loss: 1.3019 - val_acc: 0.5365
Epoch 5/100
1250/1250 [==============================] - 31s 25ms/step - loss: 1.3716 - acc: 0.5061 - val_loss: 1.2687 - val_acc: 0.5507
Epoch 6/100
1250/1250 [==============================] - 32s 25ms/step - loss: 1.3200 - acc: 0.5258 - val_loss: 1.1982 - val_acc: 0.5708
Epoch 7/100
1250/1250 [==============================] - 31s 25ms/step - loss: 1.2777 - acc: 0.5439 - val_loss: 1.1538 - val_acc: 0.5920
Epoch 8/100
1250/1250 [==================

The prediction accuracy on validation data and test data are 0.7749 and 0.7721, respectively. The difference is smaller than the simple hold-out validation, which is 0.7658 and 0.7608, respectively.

In [0]:
from keras import backend as K 

K.clear_session()